packages/stack/compiler/src/index.ts

```diff
@@ -17,53 +21,131 @@ class Compiler {
     embark.events.setCommandHandler("compiler:contracts:compile", this.compile_contracts.bind(this));
   }
 
-  private compile_contracts(contractFiles: any[], cb: any) {
+  private compile_contracts(contractFiles: any[], cb: Callback<any>) {
     if (contractFiles.length === 0) {
       return cb(null, {});
     }
 
-    const compiledObject: {[index: string]: any} = {};
+    async.waterfall(
+      [
+        (next: Callback<any[]>) => {
+          this.plugins.runActionsForEvent("compiler:contracts:compile:before", contractFiles, (err?: Error | null, contractFilesArr: any[] = []) => {
+            if (err) {
+              return next(err);
+            }
 
-    const compilerOptions = {
-      isCoverage: this.isCoverage,
-    };
+            const _contractFilesArr: any[] = [];
+            const _dappPath = dappPath();
+            const osTmpDir = os.tmpdir();
 
-    async.eachObject(this.getAvailableCompilers(),
-      (extension: string, compilers: any, next: any) => {
-        const matchingFiles = contractFiles.filter(this.filesMatchingExtension(extension));
-        if (matchingFiles.length === 0) {
-          return next();
-        }
+            for (const file of contractFilesArr) {
+              if (file instanceof File) {
+                _contractFilesArr.push(file);
+                continue;
+              }
 
-        async.someLimit(compilers, 1, (compiler: any, someCb: Callback<boolean>) => {
-          compiler.call(compiler, matchingFiles, compilerOptions, (err: any, compileResult: any) => {
-            if (err) {
-              return someCb(err);
+              if (!file.path) {
+                err = new TypeError("path property was missing on contract file object");
+                break;
+              }
+
+              let filePath: string;
+              if (!path.isAbsolute(file.path)) {
+                filePath = dappPath(file.path);
+              } else {
+                filePath = path.normalize(file.path);
+              }
+
+              if (![_dappPath, osTmpDir].some((dir) => filePath.startsWith(dir))) {
+                err = new Error("path must be within the DApp project or the OS temp directory");
+                break;
+              }
+
+              _contractFilesArr.push(
+                new File({
+                  originalPath: file.path,
+                  path: file.path,
+                  resolver: (callback: Callback<any>) => {
+                    fs.readFile(file.path, { encoding: "utf-8" }, callback);
+                  },
+                  type: Types.dappFile,
+                }),
+              );
             }
-            if (compileResult === false) {
-              // Compiler not compatible, trying the next one
-              return someCb(null, false);
+
+            if (err) {
+              return next(err);
             }
-            Object.assign(compiledObject, compileResult);
-            someCb(null, true);
+
+            next(null, _contractFilesArr);
```

Comment:
This can be simplified to:

```javascript
next(err, _contractFilesArr);
```

Reply:
Done. Though I guarded against the possibility of passing `false` with `err || null` which in the case of falsy `err` will be `null`.